crates/ruff/src/rules/flake8_pyi/rules/unnecessary_literal_union.rs

```diff
@@ -0,0 +1,119 @@
+use ruff_python_semantic::SemanticModel;
+use rustpython_parser::ast::{self, Expr, Operator, Ranged};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for the presence of multiple literal types in a union.
+///
+/// ## Why is this bad?
+/// Literal types accept multiple arguments and it is clearer to specify them
+/// as a single literal.
+///
+/// ## Example
+/// ```python
+/// from typing import Literal
+///
+/// field: Literal[1] | Literal[2]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from typing import Literal
+///
+/// field: Literal[1, 2]
+/// ```
+#[violation]
+pub struct UnnecessaryLiteralUnion {
+    members: Vec<String>,
+}
+
+impl Violation for UnnecessaryLiteralUnion {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!(
+            "Multiple literal members in a union. Use a single literal, e.g. `Literal[{}]`",
+            self.members.join(", ")
+        )
+    }
+}
+
+/// PYI030
+pub(crate) fn unnecessary_literal_union<'a>(checker: &mut Checker, expr: &'a Expr) {
+    let mut literal_exprs = Vec::new();
```

Comment:
We may want to consider using a `SmallVec` e.g. with a size of `1` to avoid allocating a `Vec` on the heap for all *proper* literal union usages

https://docs.rs/smallvec/latest/smallvec/struct.SmallVec.html